src/v/cluster/tests/leader_balancer_test.cc

```diff
@@ -59,6 +87,299 @@ BOOST_AUTO_TEST_CASE(greedy) {
+    BOOST_REQUIRE(no_movement({
+      {{1, 2, 3}, {}},
+    }));
+
+    BOOST_REQUIRE(no_movement({
+      {{1}, {2}},
+      {{2}, {1}},
+    }));
+
+    // not exactly balanced, but as balanced as possible,
+    // i.e., lead loaded node has 1 less than the most loaded node
+    BOOST_REQUIRE(no_movement({
+      {{1, 3}, {2}},
+      {{2}, {1, 3}},
+    }));
+}
+
+BOOST_AUTO_TEST_CASE(greedy_obeys_replica) {
+    // unbalanced, but the overloaded node has no replicas it can
+    // send its groups to
+    BOOST_REQUIRE(no_movement({{{1, 2, 3}, {}}, {{4}, {}}, {{}, {4}}}));
+}
+
+bool operator==(const reassignment& l, const reassignment& r) {
+    return l.group == r.group && l.from == r.from && l.to == r.to;
+}
+
+ss::sstring to_string(const reassignment& r) {
+    return fmt::format("{{group: {} from: {} to: {}}}", r.group, r.from, r.to);
+}
+
+/**
+ * @brief Test helper which checks that the given spec and other parameters
+ * result in the expected resassignment.
+ */
+ss::sstring expect_movement(
+  const cluster_spec& spec,
+  const reassignment& expected_reassignment,
+  const absl::flat_hash_set<int>& muted = {},
+  const absl::flat_hash_set<int>& skip = {}) {
+    auto [index, balancer] = from_spec(spec, muted);
+
+    absl::flat_hash_set<raft::group_id> skip_typed;
+    std::transform(
+      skip.begin(),
+      skip.end(),
+      std::inserter(skip_typed, skip_typed.begin()),
+      [](auto groupid) { return raft::group_id{groupid}; });
+
+    auto reassignment = balancer.find_movement(skip_typed);
+
+    if (!reassignment) {
+        return "no reassignment occurred";
+    }
+
+    check_valid(index, *reassignment);
+
+    if (!(*reassignment == expected_reassignment)) {
+        return fmt::format(
+          "Reassignment not as expected.\nExpected: {}\nActual:   {}\n",
+          to_string(expected_reassignment),
+          to_string(*reassignment));
+    }
+
+    return "";
+}
+
+/**
+ * @brief Helper to create a reassignment from group, from and to passed as
+ * integers.
+ */
+static reassignment re(unsigned group, int from, int to) {
+    return {
+      raft::group_id(group),
+      model::broker_shard{model::node_id(from)},
+      model::broker_shard{model::node_id(to)}};
+}
+
+BOOST_AUTO_TEST_CASE(greedy_simple_movement) {
+    // simple balancing from {2, 0} leader counts to {1, 1}
+    BOOST_REQUIRE_EQUAL(
+      expect_movement(
+        {
+          // clang-format off
+          {{1, 2}, {}},
+          {{}, {1, 2}}
+          // clang-format on
+        },
+        re(1, 0, 1)),
+      "");
+}
+
+BOOST_AUTO_TEST_CASE(greedy_highest_to_lowest) {
+    // balancing should occur from the most to least loaded node if that is
+    // possible
+    BOOST_REQUIRE_EQUAL(
+      expect_movement(
+        {
+          // clang-format off
+            {{1, 2},    {-1}},
+            {{3, 4, 5}, {-1}},                  // from 1
+            {{6, 7},    {-1}},
+            {{},        {1, 2, 3, 4, 5, 6, 7}}  // to 3
```

Comment:
The way to read these specs is:

1. Each line is a node, with two lists: leader and follower groups for the node.
2. So the first line here (node 0) means node 0 is the leader for groups 1 and 2, plus follower of -1, where -1 just means "follow everything I'm not a leader of", in this case groups 3 through 7.